### PR TITLE
Cpp loader

### PR DESCRIPTION
**Pull request by localcc**

### CODE REVIEW

File: dll_injector/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
+use std::{
+    mem::{size_of, transmute},
+    ptr,
+};
+
+use error::InjectorError;
```
Comment (konsti219) [inline dll_injector/src/lib.rs:6]: Follow preferred import order please.

File: dll_injector/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+pub mod error;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Process {
+    pid: u32,
+}
+
+impl Process {
+    pub fn find_process(starts_with: &str) -> Result<Option<Self>, InjectorError> {
```
Comment (konsti219) [inline dll_injector/src/lib.rs:38]: Maybe extend the Error type for this?

File: dll_injector/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                    }));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    pub fn wait_for_process(starts_with: &str) -> Result<Self, InjectorError> {
+        let mut process = Process::find_process(starts_with)?;
```
Comment (konsti219) [inline dll_injector/src/lib.rs:59]: Set this by default to None to avoid code duplication.

File: dll_injector/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        while let None = process {
+            process = Process::find_process(starts_with)?;
+        }
+
+        Ok(process.unwrap())
```
Comment (konsti219) [inline dll_injector/src/lib.rs:64]: Maybe `.ok_or` would be more idiomatic? Depends also on Error type changes.

File: dll_injector/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    fn foreach_thread(&self, cb: fn(HANDLE)) -> Result<(), InjectorError> {
```
Comment (konsti219) [inline dll_injector/src/lib.rs:67]: ~~What is `cb`?~~
Just call it `callback`?

File: github_helpers/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,56 @@
+use std::{error::Error, fs::File, io};
```
Comment (konsti219) [inline github_helpers/src/lib.rs:1]: import style

File: unreal_modloader/Cargo.toml [hunk 7]
```diff
@@ -35,6 +39,8 @@ semver.workspace = true
 serde.workspace = true
 serde_json.workspace = true
 steamlocate = "1.1.0"
+include-flate = "0.1.4"
```
Comment (konsti219) [inline unreal_modloader/Cargo.toml:42]: Is this dep actually used rn?

File: unreal_modloader/src/app.rs [hunk 8]
```diff
@@ -15,12 +15,17 @@ use log::{debug, info};
 use parking_lot::Mutex;
 use semver::Version;
 
-use crate::error::ModLoaderWarning;
 use crate::game_mod::{GameMod, SelectedVersion};
 use crate::mod_processing::dependencies::DependencyGraph;
 use crate::update_info::UpdateInfo;
 use crate::FileToProcess;
 use crate::{background_work::BackgroundThreadMessage, error::ModLoaderError};
+use crate::{error::ModLoaderWarning, ModLoaderAppData};
```
Comment (konsti219) [inline unreal_modloader/src/app.rs:23]: import style

File: unreal_modloader/src/config.rs [hunk 9]
```diff
@@ -17,12 +17,28 @@ pub struct IconData {
     pub height: u32,
 }
 
-pub trait InstallManager: Debug + std::marker::Send {
```
Comment (konsti219) [inline unreal_modloader/src/config.rs:20]: this is... acceptable

File: unreal_modloader/src/app.rs [hunk 10]
```diff
@@ -544,19 +549,54 @@ impl ModLoaderApp {
...
-                                };
-                            }
+                            self.launch_game(&mut data);
                         }
                     });
                 });
             });
     }
 
+    fn launch_game(&self, data: &mut ModLoaderAppData) {
```
Comment (konsti219) [inline unreal_modloader/src/app.rs:559]: This probably should be moved out of the UI file.

**Review by konsti219 — COMMENTED** [review]
Looking good so far

File: dll_injector/src/lib.rs (hunk 3 above)
Comment (localcc) [reply dll_injector/src/lib.rs:59]: I'm not sure what you mean

Comment (konsti219) [reply dll_injector/src/lib.rs:59]: `let mut process = None;`

Just write find_process once

File: unreal_modloader/src/app.rs (hunk 10 above)
Comment (konsti219) [reply unreal_modloader/src/app.rs:559]: Sounds good

File: dll_injector/src/lib.rs (hunk 4 above)
Comment (localcc) [reply dll_injector/src/lib.rs:63]: There isn't a case where we get out of the loop and still have the process set to None

**Review by konsti219 — APPROVED** [review]

**Review by konsti219 — APPROVED** [review]

**Pull request merged**